sdk/nodejs/output.ts

```diff
@@ -445,21 +433,88 @@ export function output<T>(val: Input<T | undefined>): Output<Unwrap<T | undefine
+            isOutput: true,
+         };
     }
     else if (val instanceof Array) {
-        return <any>all(val.map(output));
+        const allValues = [];
+        let hasOutputs = false;
+        for (const v of val) {
+            const ev = outputRec(v);
+
+            allValues.push(ev.value);
+            if (ev.isOutput) {
+                hasOutputs = true;
+            }
+        }
+
+        if (!hasOutputs) {
+            return { value: allValues };
+        }
+
+        const promisedArray = Promise.all(allValues.map(v => getAwaitableValue(v)));
+        const [syncResources, isKnown, isSecret, allResources] = getResourcesAndDetails(allValues);
+        return {
+            value: (new Output<any>(syncResources, promisedArray, isKnown, isSecret, allResources)),
+            isOutput: true,
+        };
     }
     else {
-        const unwrappedObject: any = {};
+        const promisedValues: { key: string; value: any }[] = [];
+        let hasOutputs = false;
         Object.keys(val).forEach(k => {
-            unwrappedObject[k] = output((<any>val)[k]);
+            const ev = outputRec(val[k]);
+            promisedValues.push({key: k, value: ev.value });
+
+            if (ev.isOutput) {
+                hasOutputs = true;
+            }
         });
 
-        return <any>all(unwrappedObject);
+        if (!hasOutputs) {
+            return { value: promisedValues.reduce((o, kvp) => { o[kvp.key] = kvp.value; return o; }, <any>{}) };
+        }
+
+        const promisedObject = getPromisedObject(promisedValues);
+        const [syncResources, isKnown, isSecret, allResources] = getResourcesAndDetails(promisedValues.map(kvp => kvp.value));
+        return { value: new Output<any>(syncResources, promisedObject, isKnown, isSecret, allResources), isOutput: true };
+    }
+}
+
+/**
+ * [output] takes any Input value and converts it into an Output, deeply unwrapping nested Input
+ * values as necessary.
+ *
+ * The expected way to use this function is like so:
+ *
+ * ```ts
+ *      var transformed = pulumi.output(someVal).apply(unwrapped => {
+ *          // Do whatever you want now.  'unwrapped' will contain no outputs/promises inside
+ *          // here, so you can easily do whatever sort of transformation is most convenient.
+ *      });
+ *
+ *      // the result can be passed to another Resource.  The dependency information will be
+ *      // properly maintained.
+ *      var someResource = new SomeResource(name, { data: transformed ... });
+ * ```
+ */
+export function output<T>(val: Input<T>): Output<Unwrap<T>>;
+export function output<T>(val: Input<T> | undefined): Output<Unwrap<T | undefined>>;
+export function output<T>(val: Input<T | undefined>): Output<Unwrap<T | undefined>> {
+    const ov = outputRec(val);
+    if (!ov.isOutput) {
+        return createSimpleOutput(ov.value);
+    }
+
+    // ov.value must already be an output.
+    if (!Output.isInstance(ov.value)) {
```

Comment:
this just seems strange to me.  if `isOutput` is true, then we have to have an Output.  So whynot check for that? i'm missing somethin here.